proposals/003-installer/installer-guidelines.md

```diff
@@ -0,0 +1,79 @@
+should be short and easy, and it should yield an environment that is easy to use with the supporting
+literature. In addition, building simple programmes with a fresh installation should be as quick
+and painless as possible.
+
+These guidelines will set out the criteria that the Haskell Foundation will use in evaluating
+installers. We do not expect any individual to meet all these criteria initially but we would like
+them to be met by a collection of installers as quickly as possible.
+
+## 1. Cross Platform Coverage
+
+As many platforms should be covered by the installers as possible. For now, the minimum platforms
+that should be covered should include 64-bit Windows, macOS and Linux running on
+Intel-x86-compatible processors. This set of minimal platforms will soon have to be extended to
+include ARM-architecture processors.
+
+## 2. Uniformity of Delivered Environment
+
+Each installer should deliver a Haskell development environment that includes the following.
+
+  * The current recommended GHC toolchain, as determined by the installer, should be installed out
+    of the box, ready to use by all of the installed build systems.
+
+  * Both of the principal build systems (`cabal-install` and `stack`) should be installed,
+    available and ready for use in the way their associated development communities would expect.
+    (If an installer interrogates the user on installation about which components to install it is
+    fine to make `cabal-install` or `stack` optional, provided are both installed when the default
+    options are take.)
+
+  * It should be possible to install and manage multiple toolchains that both `stack` and
+    `cabal-install` can use. This requires that both `stack` and `cabal-install` need to be
+    configurable to access a GHC toolchain that has been installed by the other and the installer
+    must configure both toolchains to work with the `starter` GHC toolchain.
+
+## 3. User Installable
+
+It should be possible to install Haskell in the user environment without recourse to administrator
+privileges.
+
+## 4. Editor/IDE Support
+
+Although not an initial requirement, We would like to see support for installing important,
+development infrastructure beyond the toolchains and the build systems, such as the Haskell Language
+Server or the `haskell-mode` for an `emacs` installation.  The installer might not carry out any
+such installation, but it could point the user at supporting documentation.
+
+## 5. Preinstalled Packages
+
+This won't be a minimum requirement to start with but the Haskell Foundation would like to encourage
+Haskell installation systems to install a core set of precompiled packages, apart from `base`, when
+installing the default compiler toolchain (at least). The objective is to allow developers
+(especially new developers) to be able to build certain kinds of simple applications quickly and
+painlessly after a fresh installation.
```

Comment:
Not clear what this means. GHC already ships with core libraries. Installing more libraries makes little sense for both stack and cabal-install, because they are project-centric.

Reply:
I do see the point in this, mainly because I think the `ghci` use case is important.  But since the libraries to be available aren't really specified, it is hard to evaluate the requirement today.  In any case, it's easy enough to use `cabal` to install new libraries from Hackage into the default environment.

Reply:
Yes, but libraries are clearly out of scope for installers, imo. And default environment is a semi broken thing in cabal.

Reply:
I think it would be reasonable to say that "install a Haskell development environment" includes making certain libraries available.  Indeed, it does do exactly that, but currently those libraries are mostly dictated by what GHC needs, rather than what users need.  Cabal and stack don't need this because they solve (and/or read-from-config, which is a trivial kind of solving) the version constraints for a project in their own way, and install libraries the way they want.  But if the non-project GHCi use case is important (and I definitely think it is), then it would make sense to ask that certain libraries are available by default in a newly installed Haskell development environment.  That makes it a concern when evaluating installers.

That doesn't mean we have to land on a technical solution where tools like ghcup contain the implementation for this.  It might be that the right way to accomplish the goal is to include the right libraries by default in a new GHC installation.  Maybe the right libraries are _already_ included with GHC.  We cannot know unless we have a list of those libraries.  Prior art for this question has to include the Haskell Platform, which had a list of libraries that it made available by default.

Reply:
> But if the non-project GHCi use case is important (and I definitely think it is), then it would make sense to ask that certain libraries are available by default in a newly installed Haskell development environment. That makes it a concern when evaluating installers.

Installers aren't a substitute for good documentation. Users need to figure out how to install/use libraries *anyway*. 

The few times where users didn't know how to do that, we were able to help them quickly on IRC.

So:

1. this is mainly a documentation issue
2. cabal and stack need to provide better interfaces to start up an environment with X libraries
3. installer tools can point to the documentation at the end or provide short instructions as is already done: https://gitlab.haskell.org/haskell/ghcup-hs/-/blob/bf74d1e828280a0efc86816aaea82df0b302fa82/bootstrap-haskell#L115

Reply:
As a sidenote: I really believe cabal sandboxes were the right interface for this plug-and-play ghci story. But they were (sadly) removed. Today you can also pass the `--store-dir` argument, but that's overall not good UX.